src/main/java/com/aws/greengrass/secretmanager/SecretManager.java

```diff
@@ -181,11 +198,25 @@ private AWSSecretResponse fetchAndEncryptAWSResponse(GetSecretValueRequest reque
                 .build();
     }
 
+    void waitForInitialization() throws SecretCryptoException, InterruptedException {
+        synchronized (initialized) {
+            while (!initialized.isSet()) {
+                initialized.wait();
+            }
+        }
+        if (initialized.isSet()) {
+            if (initialized.getValue() != null) {
+                throw initialized.getValue();
+            }
```

Comment:
In which case do you need to throw here. Wouldn't you just throw from the initialization block, because if we cannot init, then that is fatal, Secret manager does not come up.

Reply:
Initialization block is running in a random thread, we can't throw to that because no one is listening. We cannot block inside the constructor waiting for the security service to startup. All constructors are run on a single thread, so if we block in the constructor waiting for the security service, it could deadlock because the security service was scheduled after us.

Reply:
Ok. Are you sure, that you do not need to lock the write? I think this is the first time I am seeing that read is synchronized but write is not. I am not sure if this is enough to avoid memory inconsistency errors.

Reply:
exactly, can't wait for security service in the constructor. DI is single threaded.

Reply:
This isn't synchronizing the read at all; in order to `wait` we need to be synchronized on the object that we're waiting for. Inside the `set` call it synchronizes in order to call `notifyAll`